### PR TITLE
#19361 Rewrite Duplicator/Duplicator2 using GraphStage instead of StatefulStage

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

**Comment by ktoso** [comment]
OK TO TEST

File: akka-docs-dev/rst/scala/code/docs/stream/FlowStagesSpec.scala [hunk 1]
```diff
@@ -107,18 +106,33 @@ class FlowStagesSpec extends AkkaSpec with ScalaFutures {
...
-      class Duplicator[A]() extends StatefulStage[A, A] {
-        override def initial: StageState[A, A] = new StageState[A, A] {
-          override def onPush(elem: A, ctx: Context[A]): SyncDirective =
-            emit(List(elem, elem).iterator, ctx)
-        }
+      class Duplicator[A] extends GraphStage[FlowShape[A, A]] {
+        val in = Inlet[A]("Duplicator.in")
+        val out = Outlet[A]("Duplicator.out")
+
+        val shape: FlowShape[A, A] = FlowShape.of(in, out)
```
Comment (drewhk) [inline akka-docs-dev/rst/scala/code/docs/stream/FlowStagesSpec.scala]: this is scala, so the simple apply is preferred.

**Comment by drewhk** [comment]
LGTM

File: akka-samples/akka-docs-java-lambda/src/test/java/docs/stream/FlowStagesDocTest.java [hunk 2]
```diff
@@ -141,11 +141,27 @@ public Filter2(Predicate<A> p) {
...
+        {
+          setHandler(in, new AbstractInHandler() {
+            @Override public void onPush() {
+              A elem = grab(in);
+              emitMultiple(out, Arrays.asList(elem, elem).iterator());
+            }
+          });
+          setHandler(out, new AbstractOutHandler() {
+            @Override public void onPull() { pull(in); }
+          });
```
Comment (ktoso) [inline akka-samples/akka-docs-java-lambda/src/test/java/docs/stream/FlowStagesDocTest.java]: There is another one here, 
`class Duplicator<A> extends PushPullStage<A, A> {`

Can be addressed in a separate PR though

**Comment by ktoso** [comment]
LGTM after `of(...)` changed to `(...)`, in examples we want to show idiomatic use.

File: akka-samples/akka-docs-java-lambda/src/test/java/docs/stream/FlowStagesDocTest.java (hunk 2 above)
Comment (talp-traiana) [inline akka-samples/akka-docs-java-lambda/src/test/java/docs/stream/FlowStagesDocTest.java]: I don't mind doing it in this PR, I just wasn't sure if you wanted it included - the PushPullStage isn't deprecated yet, and there are places (such as FlowStagesSpec.scala) where the Duplicator is part of a test demonstrating PushPullStages specifically.

Comment (ktoso) [inline akka-samples/akka-docs-java-lambda/src/test/java/docs/stream/FlowStagesDocTest.java]: Let's do it separately. We're moving streams to the master branch today, so let's get this one merged ASAP :)

**Comment by talp-traiana** [comment]
OK, done. Good luck with the move ^_^

**Comment by ktoso** [comment]
LGTM, thanks!

**Pull request merged**